### ISSUE TITLE
cli command update-user-pool failure with Unknown parameter in LambdaConfig

### ISSUE THREAD

**Issue by rvairavana** [aws/aws-cli#4946]
Hi
[updateUserPool_stage_success.txt](https://github.com/aws/aws-cli/files/4188571/updateUserPool_stage_success.txt)
[updateUserPool_error.txt](https://github.com/aws/aws-cli/files/4188572/updateUserPool_error.txt)

we are trying to associate the cognito user-pool with new cognito lambda ARNs, and its failing on one environment. We dont see this issue on any other environment. all environment runs on the same CLI version and user permissions. but fails only one one env, but not others.

CLI command:

aws cognito-idp update-user-pool --user-pool-id <poolid> --policies '{
                          "PasswordPolicy": {
                            "MinimumLength": 6,
                            "RequireUppercase": false,
                            "RequireLowercase": false,
                            "RequireNumbers": false,
                            "RequireSymbols": false
                          }
                        }' --email-configuration '{
                          "SourceArn": "arn:aws:ses:us-east-1:<>:identity/customerservice@cs.shutterfly.com",
                          "ReplyToEmailAddress": "noreply@cs.shutterfly.com "
                        }' --lambda-config '{
                          "PreSignUp": "arn:aws:lambda:us-east-1:<accountid>:function:sfly-aws-prod-cognito_pre_sign_up_validation",
                          "CustomMessage": "arn:aws:lambda:us-east-1:<accountid>:function:sfly-aws-prod-cognito_custom_message",
                          "PostConfirmation": "arn:aws:lambda:us-east-1:<accountid>:function:sfly-aws-prod-cognito_post_confirmation",
                          "DefineAuthChallenge": "arn:aws:lambda:us-east-1:<accountid>:function:sfly-aws-prod-cognito_define_auth_challenge",
                          "CreateAuthChallenge": "arn:aws:lambda:us-east-1:<accountid>:function:sfly-aws-prod-cognito_create_auth_challenge",
                          "VerifyAuthChallengeResponse": "arn:aws:lambda:us-east-1:<accountid>:function:sfly-aws-prod-cognito_verify_auth_challenge_response",
                          "PreTokenGeneration": "arn:aws:lambda:us-east-1:<accountid>:function:sfly-aws-prod-cognito_pre_token_generation",
                          "UserMigration": "arn:aws:lambda:us-east-1:<accountid>:function:sfly-aws-prod-cognito_user_migration"
                        }' --region us-east-1

ERROR:
Parameter validation failed:
Unknown parameter in LambdaConfig: "PreTokenGeneration", must be one of: PreSignUp, CustomMessage, PostConfirmation, PreAuthentication, PostAuthentication, DefineAuthChallenge, CreateAuthChallenge, VerifyAuthChallengeResponse
Unknown parameter in LambdaConfig: "UserMigration", must be one of: PreSignUp, CustomMessage, PostConfirmation, PreAuthentication, PostAuthentication, DefineAuthChallenge, CreateAuthChallenge, VerifyAuthChallengeResponse

opened a aws support ticket with aws (6796241221), and aws support analyzed the issue and as per their suggestion, opening this issue ticket.

one difference we noticed from the debug log is the json data files are retrieved files from different directories:
where it worked:
2020-02-10 13:15:51,824 - MainThread - botocore.loaders - DEBUG - Loading JSON file: /home/distro/.aws/models/cognito-idp/2016-04-18/service-2.json

where it failed:
2020-02-10 21:56:00,475 - MainThread - botocore.loaders - DEBUG - Loading JSON file: /usr/lib/python2.7/site-packages/botocore/data/cognito-idp/2016-04-18/service-2.json

we couldn't figure out any difference between the env where the exact same command worked vs. failed.

attached both debug logs (one succeeded and the other one that failed for comparison)
Thanks,
Rajeev

**Comment by joguSD** [comment]
This line:
```
/home/distro/.aws/models/cognito-idp/2016-04-18/service-2.json
```

Indicates a custom model for the service that is independent of the CLI version. This model doesn't have the parameter that you're trying to pass hence why it's failing. Unless you have a specific reason for using the custom models I would suggest removing them from `~/.aws/models`.

**Comment by rvairavana** [comment]
The model you mentioned above (/home/distro/.aws/models) is in fact working fine.

Where its not working, we noticed that the directory is different:

2020-02-10 21:56:00,475 - MainThread - botocore.loaders - DEBUG - Loading JSON file: /usr/lib/python2.7/site-packages/botocore/data/cognito-idp/2016-04-18/service-2.json

But we still do not know why the botocore.loaders are different. everything is supposed to be exactly the same on both environments.

**Comment by joguSD** [comment]
@rvairavana The failing logs show a rather old CLI version that doesn't have the parameter yet. I would recommend updating the CLI version and removing the custom model.

**Comment by rvairavana** [comment]
thanks, from where you find the cli version?

from the debug logs i attached, both (failed and succeeded logs) have the exact same cli version.

from failed log - 
2020-02-10 21:56:00,472 - MainThread - awscli.clidriver - DEBUG - CLI version: aws-cli/1.11.63 Python/2.7.5 Linux/3.10.0-514.26.1.el7.x86_64 botocore/1.5.95

from succeeded log -
2020-02-10 13:15:51,820 - MainThread - awscli.clidriver - DEBUG - CLI version: aws-cli/1.11.63 Python/2.7.5 Linux/3.10.0-514.26.1.el7.x86_64 botocore/1.5.95

we can try update the CLI version, but with the same old cli version the command worked fine on all 4 test environments.

**Comment by rvairavana** [comment]
The issue got resolved after upgrading to aws-cli/1.16.48. But it still a mystery how it worked in few other environment with the same old version.
But we can close this issue. thanks
